Foundation/NSURL.swift

```diff
@@ -638,13 +665,462 @@ open class NSURL : NSObject, NSSecureCoding, NSCopying {
+                newValues[.isSystemImmutableKey] = try hasFlag(path: path, flag: FS_IMMUTABLE_FL)
+                #else
+                NSUnsupported()
+                #endif
+                
+                
+            case .isHiddenKey:
+                #if canImport(Darwin) || os(FreeBSD)
+                newValues[.isHiddenKey] = stat.st_flags & UInt32(UF_HIDDEN) != 0
+                #elseif os(Windows)
+                newValues[.isHiddenKey] = try hasFlag(path: path, flag: FILE_ATTRIBUTE_HIDDEN)
+                #else
+                newValues[.isHiddenKey] = lastPathComponent?.hasPrefix(".")
+                #endif
+                
+            case .isReadableKey:
+                newValues[.isReadableKey] = FileManager.default.isReadableFile(atPath: path)
+                
+            case .isWritableKey:
+                newValues[.isWritableKey] = FileManager.default.isWritableFile(atPath: path)
+                
+            case .isExecutableKey:
+                newValues[.isExecutableKey] = FileManager.default.isExecutableFile(atPath: path)
+                
+            case .generationIdentifierKey, .documentIdentifierKey:
+                fallthrough
+                
+            case .addedToDirectoryDateKey:
+                fallthrough
+                
+            // These keys are Darwin specific. Some can be implemented using xattrs, but still no
+            // linux disto would honor them!
+            case .hasHiddenExtensionKey, .isPackageKey , .isApplicationKey, .typeIdentifierKey,
+                 .localizedTypeDescriptionKey, .labelNumberKey, .labelColorKey, .localizedLabelKey,
+                 .effectiveIconKey, .customIconKey, .fileSecurityKey, .isExcludedFromBackupKey,
+                 .tagNamesKey, .isMountTriggerKey, .fileResourceIdentifierKey, .quarantinePropertiesKey,
+                 .thumbnailDictionaryKey, .thumbnailKey, .isAliasFileKey:
+                NSUnsupported()
+                
+             case .volumeTotalCapacityKey, .volumeAvailableCapacityKey, .volumeResourceCountKey,
+                  .volumeAvailableCapacityForImportantUsageKey, .volumeAvailableCapacityForOpportunisticUsageKey:
+                let attributes = try? FileManager.default.attributesOfFileSystem(forPath: path)
+                newValues[.volumeTotalCapacityKey] = attributes?[.systemSize]
+                newValues[.volumeAvailableCapacityKey] = attributes?[.systemFreeSize]
+                newValues[.volumeAvailableCapacityForImportantUsageKey] = attributes?[.systemFreeSize]
+                newValues[.volumeAvailableCapacityForOpportunisticUsageKey] = attributes?[.systemFreeSize]
+                newValues[.volumeResourceCountKey] = attributes?[.systemNodes]
+                
+            case .volumeURLKey, .isVolumeKey:
+                let rootPath: String
+                #if canImport(Darwin) || os(FreeBSD)
+                var root = try _statfs(path: path).f_mntonname
```

Comment:
It looks like `_statfs(path: path)` is called in multiple places so it might be best to move it outside of the loop and only call it once (and only if needed).

The same applies to `mountedVolumeURLS` as it could be called twice.

Reply:
There are only two places which I use `_statfs()` in this loop. Furthermore it is only available in Darwin and FreeBSD thus moving this would add far more complexity to code.